File: decompiled-src-verify/io/legado/app/utils/NetworkUtils.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  kotlin.Lazy
 *  kotlin.LazyKt
 *  kotlin.Metadata
 *  kotlin.jvm.functions.Function0
 *  kotlin.jvm.internal.Intrinsics
 *  kotlin.text.StringsKt
 *  okhttp3.HttpUrl
 *  okhttp3.Request
 *  okhttp3.Response
 *  org.jetbrains.annotations.NotNull
 *  org.jetbrains.annotations.Nullable
 *  retrofit2.Response
 */
package io.legado.app.utils;

import io.legado.app.utils.NetworkUtils;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.net.URL;
import java.util.BitSet;
import java.util.Enumeration;
import java.util.regex.Pattern;
import kotlin.Lazy;
import kotlin.LazyKt;
import kotlin.Metadata;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.internal.Intrinsics;
import kotlin.text.StringsKt;
import okhttp3.HttpUrl;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Metadata(mv={1, 5, 1}, k=1, xi=48, d1={"\u0000H\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0010\u000e\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0004\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000b\n\u0002\b\u0003\n\u0002\u0010\f\n\u0002\b\u0003\b\u00c6\u0002\u0018\u00002\u00020\u0001B\u0007\b\u0002\u00a2\u0006\u0002\u0010\u0002J\u0018\u0010\f\u001a\u00020\r2\b\u0010\u000e\u001a\u0004\u0018\u00010\u000f2\u0006\u0010\u0010\u001a\u00020\rJ\u0018\u0010\f\u001a\u00020\r2\b\u0010\u000e\u001a\u0004\u0018\u00010\r2\u0006\u0010\u0010\u001a\u00020\rJ\u0012\u0010\u0011\u001a\u0004\u0018\u00010\r2\b\u0010\u0012\u001a\u0004\u0018\u00010\rJ\b\u0010\u0013\u001a\u0004\u0018\u00010\u0014J\u0010\u0010\u0015\u001a\u00020\r2\b\u0010\u0012\u001a\u0004\u0018\u00010\rJ\u0012\u0010\u0016\u001a\u00020\r2\n\u0010\u0017\u001a\u0006\u0012\u0002\b\u00030\u0018J\u000e\u0010\u0019\u001a\u00020\u001a2\u0006\u0010\u001b\u001a\u00020\rJ\u0010\u0010\u001c\u001a\u00020\u001a2\u0006\u0010\u001d\u001a\u00020\u001eH\u0002J\u000e\u0010\u001f\u001a\u00020\u001a2\u0006\u0010 \u001a\u00020\rR\u0016\u0010\u0003\u001a\n \u0005*\u0004\u0018\u00010\u00040\u0004X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u001b\u0010\u0006\u001a\u00020\u00078BX\u0082\u0084\u0002\u00a2\u0006\f\n\u0004\b\n\u0010\u000b\u001a\u0004\b\b\u0010\t\u00a8\u0006!"}, d2={"Lio/legado/app/utils/NetworkUtils;", "", "()V", "IPV4_PATTERN", "Ljava/util/regex/Pattern;", "kotlin.jvm.PlatformType", "notNeedEncoding", "Ljava/util/BitSet;", "getNotNeedEncoding", "()Ljava/util/BitSet;", "notNeedEncoding$delegate", "Lkotlin/Lazy;", "getAbsoluteURL", "", "baseURL", "Ljava/net/URL;", "relativePath", "getBaseUrl", "url", "getLocalIPAddress", "Ljava/net/InetAddress;", "getSubDomain", "getUrl", "response", "Lretrofit2/Response;", "hasUrlEncoded", "", "str", "isDigit16Char", "c", "", "isIPv4Address", "input", "reader-pro"})
public final class NetworkUtils {
    @NotNull
    public static final NetworkUtils INSTANCE = new NetworkUtils();
    @NotNull
    private static final Lazy notNeedEncoding$delegate = LazyKt.lazy((Function0)notNeedEncoding.2.INSTANCE);
    private static final Pattern IPV4_PATTERN = Pattern.compile("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$");

    private NetworkUtils() {
    }

    @NotNull
    public final String getUrl(@NotNull retrofit2.Response<?> response2) {
        HttpUrl httpUrl;
        Request request;
        Response networkResponse;
        Intrinsics.checkNotNullParameter(response2, (String)"response");
        Response response3 = networkResponse = response2.raw().networkResponse();
        String string = response3 == null ? null : ((request = response3.request()) == null ? null : ((httpUrl = request.url()) == null ? null : httpUrl.toString()));
        return string == null ? response2.raw().request().url().toString() : string;
    }

    private final BitSet getNotNeedEncoding() {
        Lazy lazy = notNeedEncoding$delegate;
        boolean bl = false;
        return (BitSet)lazy.getValue();
    }

    public final boolean hasUrlEncoded(@NotNull String str) {
        Intrinsics.checkNotNullParameter((Object)str, (String)"str");
        boolean needEncode = false;
        int i = 0;
        while (i < str.length()) {
            int c;
            int n = c = str.charAt(i);
            boolean bl = false;
            if (this.getNotNeedEncoding().get(n)) {
                n = i;
                i = n + 1;
                continue;
            }
            if (c == 37 && i + 2 < str.length()) {
                char c1 = str.charAt(++i);
                char c2 = str.charAt(++i);
                if (this.isDigit16Char(c1) && this.isDigit16Char(c2)) {
                    int n2 = i;
                    i = n2 + 1;
                    continue;
                }
            }
            needEncode = true;
            break;
        }
        return !needEncode;
    }

    /*
     * Enabled force condition propagation
     * Lifted jumps to return sites
     */
    private final boolean isDigit16Char(char c) {
        boolean bl;
        boolean bl2;
        if ('0' <= c) {
            if (c <= '9') {
                return true;
            }
            bl2 = false;
        } else {
            bl2 = false;
        }
        if (bl2) return true;
        if ('A' <= c) {
            if (c <= 'F') {
                return true;
            }
            bl = false;
        } else {
            bl = false;
        }
        if (bl) return true;
        if ('a' > c) return false;
        if (c > 'f') return false;
        return true;
    }

    @NotNull
    public final String getAbsoluteURL(@Nullable String baseURL, @NotNull String relativePath) {
        Intrinsics.checkNotNullParameter((Object)relativePath, (String)"relativePath");
        CharSequence charSequence = baseURL;
        boolean bl = false;
        boolean bl2 = false;
        if (charSequence == null || charSequence.length() == 0) {
            return relativePath;
        }
        charSequence = relativePath;
        bl = false;
        bl2 = false;
        if (charSequence.length() == 0) {
            return baseURL;
        }
        String relativeUrl = relativePath;
        try {
            URL absoluteUrl = new URL(StringsKt.substringBefore$default((String)baseURL, (String)",", null, (int)2, null));
            URL parseUrl = new URL(absoluteUrl, relativePath);
            String string = parseUrl.toString();
            Intrinsics.checkNotNullExpressionValue((Object)string, (String)"parseUrl.toString()");
            relativeUrl = string;
            return relativeUrl;
        }
        catch (Exception e) {
            e.printStackTrace();
            return relativeUrl;
        }
    }

    @NotNull
    public final String getAbsoluteURL(@Nullable URL baseURL, @NotNull String relativePath) {
        Intrinsics.checkNotNullParameter((Object)relativePath, (String)"relativePath");
        if (baseURL == null) {
            return relativePath;
        }
        String relativeUrl = relativePath;
        try {
            URL parseUrl = new URL(baseURL, relativePath);
            String string = parseUrl.toString();
            Intrinsics.checkNotNullExpressionValue((Object)string, (String)"parseUrl.toString()");
            relativeUrl = string;
            return relativeUrl;
        }
        catch (Exception e) {
            e.printStackTrace();
            return relativeUrl;
        }
    }

    @Nullable
    public final String getBaseUrl(@Nullable String url2) {
        String string;
        if (url2 == null || !StringsKt.startsWith$default((String)url2, (String)"http", (boolean)false, (int)2, null)) {
            return null;
        }
        int index = StringsKt.indexOf$default((CharSequence)url2, (String)"/", (int)9, (boolean)false, (int)4, null);
        if (index == -1) {
            string = url2;
        } else {
            String string2 = url2;
            int n = 0;
            boolean bl = false;
            String string3 = string2.substring(n, index);
            string = string3;
            Intrinsics.checkNotNullExpressionValue((Object)string3, (String)"(this as java.lang.Strin\u2026ing(startIndex, endIndex)");
        }
        return string;
    }

    @NotNull
    public final String getSubDomain(@Nullable String url2) {
        String string;
        String string2 = this.getBaseUrl(url2);
        if (string2 == null) {
            return "";
        }
        String baseUrl = string2;
        if (StringsKt.indexOf$default((CharSequence)baseUrl, (String)".", (int)0, (boolean)false, (int)6, null) == StringsKt.lastIndexOf$default((CharSequence)baseUrl, (String)".", (int)0, (boolean)false, (int)6, null)) {
            string2 = baseUrl;
            int n = StringsKt.lastIndexOf$default((CharSequence)baseUrl, (String)"/", (int)0, (boolean)false, (int)6, null) + 1;
            boolean bl = false;
            String string3 = string2.substring(n);
            string = string3;
            Intrinsics.checkNotNullExpressionValue((Object)string3, (String)"(this as java.lang.String).substring(startIndex)");
        } else {
            string2 = baseUrl;
            int n = StringsKt.indexOf$default((CharSequence)baseUrl, (String)".", (int)0, (boolean)false, (int)6, null) + 1;
            boolean bl = false;
            String string4 = string2.substring(n);
            string = string4;
            Intrinsics.checkNotNullExpressionValue((Object)string4, (String)"(this as java.lang.String).substring(startIndex)");
        }
        return string;
    }

    @Nullable
    public final InetAddress getLocalIPAddress() {
        Enumeration<NetworkInterface> enumeration = null;
        try {
            enumeration = NetworkInterface.getNetworkInterfaces();
        }
        catch (SocketException e) {
            e.printStackTrace();
        }
        if (enumeration != null) {
            while (enumeration.hasMoreElements()) {
                NetworkInterface nif = enumeration.nextElement();
                Enumeration<InetAddress> addresses = nif.getInetAddresses();
                if (addresses == null) continue;
                while (addresses.hasMoreElements()) {
                    InetAddress address = addresses.nextElement();
                    if (address.isLoopbackAddress()) continue;
                    String string = address.getHostAddress();
                    Intrinsics.checkNotNullExpressionValue((Object)string, (String)"address.hostAddress");
                    if (!this.isIPv4Address(string)) continue;
                    return address;
                }
            }
        }
        return null;
    }

    public final boolean isIPv4Address(@NotNull String input) {
        Intrinsics.checkNotNullParameter((Object)input, (String)"input");
        return IPV4_PATTERN.matcher(input).matches();
    }
}

File: best-of-3/src/io/legado/app/utils/NetworkUtils.kt
package io.legado.app.utils

import com.htmake.reader.synth.SyntheticContinuation
import com.htmake.reader.synth.SyntheticFunction0
import com.htmake.reader.synth.SyntheticType

import java.net.InetAddress
import java.net.NetworkInterface
import java.net.SocketException
import java.net.URL
import java.util.BitSet
import java.util.Enumeration
import java.util.regex.Pattern
import okhttp3.HttpUrl
import okhttp3.Request
import retrofit2.Response

public object NetworkUtils {
   private final val IPV4_PATTERN: Pattern =
      Pattern.compile("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$")

   private final val notNeedEncoding: BitSet by LazyKt.lazy(SyntheticFunction0.INSTANCE)
      private final get() {
         return notNeedEncoding$delegate.getValue() as BitSet;
      }

   public fun getUrl(response: Response<*>): String {
      val networkResponse: okhttp3.Response = response.raw().networkResponse();
      val var10000: java.lang.String;
      if (networkResponse == null) {
         var10000 = null;
      } else {
         val var5: Request = networkResponse.request();
         if (var5 == null) {
            var10000 = null;
         } else {
            val var6: HttpUrl = var5.url();
            var10000 = if (var6 == null) null else var6.toString();
         }
      }

      return if (var10000 == null) response.raw().request().url().toString() else var10000;
   }

   public fun hasUrlEncoded(str: String): Boolean {
      var needEncode: Boolean = false;
      var i: Int = 0;

      while (i < str.length()) {
         val c: Char = str.charAt(i);
         if (this.getNotNeedEncoding().get(c)) {
            i++;
         } else {
            if (c != '%' || i + 2 >= str.length() || !this.isDigit16Char(str.charAt(++i)) || !this.isDigit16Char(str.charAt(++i))) {
               needEncode = true;
               break;
            }

            i++;
            continue;
         }
      }

      return !needEncode;
   }

   private fun isDigit16Char(c: Char): Boolean {
      return '0' <= c && c <= '9' || 'A' <= c && c <= 'F' || 'a' <= c && c <= 'f';
   }

   public fun getAbsoluteURL(baseURL: String?, relativePath: String): String {
      if (baseURL == null || baseURL.length() == 0) {
         return relativePath;
      } else if (relativePath.length() == 0) {
         return baseURL;
      } else {
         var var9: java.lang.String = relativePath;

         try {
            val var6: java.lang.String = new URL(new URL(StringsKt.substringBefore$default(baseURL, ",", null, 2, null)), relativePath).toString();
            var9 = var6;
            return var6;
         } catch (var7: Exception) {
            var7.printStackTrace();
            return var9;
         }
      }
   }

   public fun getAbsoluteURL(baseURL: URL?, relativePath: String): String {
      if (baseURL == null) {
         return relativePath;
      } else {
         var relativeUrl: java.lang.String = relativePath;

         try {
            val var5: java.lang.String = new URL(baseURL, relativePath).toString();
            relativeUrl = var5;
            return var5;
         } catch (var6: Exception) {
            var6.printStackTrace();
            return relativeUrl;
         }
      }
   }

   public fun getBaseUrl(url: String?): String? {
      if (url != null && StringsKt.startsWith$default(url, "http", false, 2, null)) {
         val index: Int = StringsKt.indexOf$default(url, "/", 9, false, 4, null);
         val var10000: java.lang.String;
         if (index == -1) {
            var10000 = url;
         } else {
            var10000 = url.substring(0, index);
         }

         return var10000;
      } else {
         return null;
      }
   }

   public fun getSubDomain(url: String?): String {
      val var3: java.lang.String = this.getBaseUrl(url);
      if (var3 == null) {
         return "";
      } else {
         val var10000: java.lang.String;
         if (StringsKt.indexOf$default(var3, ".", 0, false, 6, null) == StringsKt.lastIndexOf$default(var3, ".", 0, false, 6, null)) {
            var10000 = var3.substring(StringsKt.lastIndexOf$default(var3, "/", 0, false, 6, null) + 1);
         } else {
            var10000 = var3.substring(StringsKt.indexOf$default(var3, ".", 0, false, 6, null) + 1);
         }

         return var10000;
      }
   }

   public fun getLocalIPAddress(): InetAddress? {
      var enumeration: Enumeration = null;

      try {
         enumeration = NetworkInterface.getNetworkInterfaces();
      } catch (var6: SocketException) {
         var6.printStackTrace();
      }

      if (enumeration != null) {
         while (enumeration.hasMoreElements()) {
            val addresses: Enumeration = (enumeration.nextElement() as NetworkInterface).getInetAddresses();
            if (addresses != null) {
               while (addresses.hasMoreElements()) {
                  val address: InetAddress = addresses.nextElement() as InetAddress;
                  if (!address.isLoopbackAddress()) {
                     val var5: java.lang.String = address.getHostAddress();
                     if (this.isIPv4Address(var5)) {
                        return address;
                     }
                  }
               }
            }
         }
      }

      return null;
   }

   public fun isIPv4Address(input: String): Boolean {
      return IPV4_PATTERN.matcher(input).matches();
   }
}
